src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -68,6 +71,56 @@ pub struct EventLoop {
     // XXX drop after everything else, just to be safe.
     /// Calloop's event loop.
     event_loop: calloop::EventLoop<'static, WinitState>,
+
+    pump_event_notifier: Option<PumpEventNotifier>,
+}
+
+#[derive(Debug)]
+struct PumpEventNotifier {
+    /// Whether we're in winit or not.
+    control: Arc<(Mutex<bool>, Condvar)>,
+    /// Thread handle.
+    _handle: JoinHandle<()>,
+}
+
+fn spawn_wakeup_thread(
+    connection: Connection,
+    awakener: calloop::ping::Ping,
+) -> Option<PumpEventNotifier> {
+    // Start from the waiting state.
+    let control = Arc::new((Mutex::new(true), Condvar::new()));
+    let control_thread = Arc::clone(&control);
+
+    let handle = std::thread::Builder::new()
+        .name(String::from("wayland pump_events wake-up notify"))
+        .spawn(move || {
+            let (lock, cvar) = &*control_thread;
+            loop {
+                let mut wait = lock.lock().unwrap();
+                while *wait {
+                    wait = cvar.wait(wait).unwrap();
+                }
+
+                if let Some(read_guard) = connection.prepare_read() {
+                    let _ = connection.flush();
```

Comment:
This needs to be a loop with polling until flush returns Ok.

Reply:
This should likely mention that, but I've never seen a wayland loop that `flush` in a loop until it works. I'd also assume that it may fail in a non-recoverable way.